File: frontend/src/api/schedules.ts
/**
 * 排期会议 API
 */
import request from './request'

export type ScheduleStatus =
  | 'pending'
  | 'confirmed'
  | 'in_progress'
  | 'completed'
  | 'cancelled'

export type ScheduleResourceType = 'instructor' | 'streamer' | 'shooting_edit' | 'other'
export type ScheduleType = 'internal_training' | 'external_salon' | 'project_live' | 'other'

export interface ScheduleConflict {
  id: number
  title: string
  start_time: string
  end_time: string
  status: string
}

export interface Schedule {
  id: number
  title: string
  schedule_type?: ScheduleType | string
  resource_type: ScheduleResourceType | string
  employee_id: number
  project_id?: number | null
  project_task_id?: number | null
  ticket_id?: number | null
  start_time: string
  end_time: string
  status: ScheduleStatus | string
  creator_id: number
  department_id?: number | null
  confirmed_by?: number | null
  confirmed_at?: string | null
  location?: string | null
  content?: string | null
  coordination_note?: string | null
  result?: string | null
  actual_hours?: number | string | null
  timesheet_id?: number | null
  feishu_sync_status?: string
  cancel_reason?: string | null
  remark?: string | null
  created_at: string
  updated_at: string
  employee_name?: string | null
  creator_name?: string | null
  confirmed_by_name?: string | null
  project_name?: string | null
  project_no?: string | null
  task_no?: string | null
  task_title?: string | null
  ticket_no?: string | null
  has_conflict?: boolean
  conflicts?: ScheduleConflict[]
  planned_hours?: number | null
}

export interface ScheduleStats {
  total: number
  pending: number
  confirmed: number
  in_progress: number
  completed: number
  cancelled: number
  conflict_count: number
  mine: number
}

export interface ResourceOption {
  id: number
  name: string
  department_name?: string | null
  job_title?: string | null
  role_names?: string[]
}

export interface PersonTreeNode {
  value: number | string
  label: string
  disabled?: boolean
  is_person?: boolean
  children?: PersonTreeNode[]
}

export interface ResourceLoad {
  employee_id: number
  employee_name: string
  resource_type: string
  planned_hours: number
  load_percent: number
  item_count: number
}

export const SCHEDULE_STATUS_LABEL: Record<string, string> = {
  pending: '待确认',
  confirmed: '已确认',
  in_progress: '进行中',
  completed: '已完成',
  cancelled: '已取消',
}

export const SCHEDULE_RESOURCE_OPTIONS = [
  { value: 'instructor', label: '讲师' },
  { value: 'streamer', label: '主播' },
  { value: 'shooting_edit', label: '拍摄剪辑' },
  { value: 'other', label: '其他' },
]

export const SCHEDULE_TYPE_OPTIONS = [
  { value: 'internal_training', label: '内部培训' },
  { value: 'external_salon', label: '外部沙龙' },
  { value: 'project_live', label: '项目直播/活动' },
  { value: 'other', label: '其他排期' },
]

export const FEISHU_SYNC_LABEL: Record<string, string> = {
  none: '未同步',
  pending: '待同步',
  synced: '已同步',
  failed: '同步失败',
}

export function fetchScheduleStats() {
  return request.get<ScheduleStats>('/schedules/stats')
}

export function fetchResourceOptions(params?: { resource_type?: string }) {
  return request.get<ResourceOption[]>('/schedules/options/resources', { params })
}

export function fetchPersonTree() {
  return request.get<PersonTreeNode[]>('/schedules/options/person-tree')
}

export function fetchResourceLoad(params: {
  resource_type: string
  date_from: string
  date_to: string
}) {
  return request.get<{ items: ResourceLoad[] }>('/schedules/resource-load', { params })
}

export function fetchSchedules(params: {
  status?: string
  resource_type?: string
  employee_id?: number
  project_id?: number
  project_task_id?: number
  date_from?: string
  date_to?: string
  scope?: string
  page?: number
  page_size?: number
}) {
  return request.get<{ total: number; items: Schedule[] }>('/schedules', { params })
}

export function fetchScheduleDetail(id: number) {
  return request.get<Schedule>(`/schedules/${id}`)
}

export function createSchedule(data: {
  title: string
  schedule_type?: string
  resource_type?: string
  employee_id: number
  project_id?: number
  project_task_id?: number
  ticket_id?: number
  start_time: string
  end_time: string
  location?: string
  content?: string
  remark?: string
}) {
  return request.post<Schedule>('/schedules', data)
}

export function updateSchedule(id: number, data: Partial<Schedule>) {
  return request.patch<Schedule>(`/schedules/${id}`, data)
}

export function confirmSchedule(id: number) {
  return request.post<Schedule>(`/schedules/${id}/confirm`)
}

export function coordinateSchedule(id: number, note: string) {
  return request.post<Schedule>(`/schedules/${id}/coordinate`, { note })
}

export function startSchedule(id: number) {
  return request.post<Schedule>(`/schedules/${id}/start`)
}

export function completeSchedule(
  id: number,
  data: { result: string; actual_hours: number; create_timesheet?: boolean },
) {
  return request.post<Schedule>(`/schedules/${id}/complete`, data)
}

export function cancelSchedule(id: number, reason?: string) {
  return request.post<Schedule>(`/schedules/${id}/cancel`, { reason })
}

File: backend/app/api/v1/schedules.py
"""排期 API。"""
from datetime import datetime
from typing import Annotated, Optional

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.api.deps import PermissionChecker
from app.database import get_db
from app.models.user import User
from app.schemas.schedule import (
    ResourceLoadListOut,
    ResourceLoadOut,
    ScheduleCancelRequest,
    ScheduleCompleteRequest,
    ScheduleCoordinateRequest,
    ScheduleCreate,
    ScheduleListOut,
    ScheduleOut,
    ScheduleStatsOut,
    ScheduleUpdate,
)
from app.services import schedule as schedule_service

router = APIRouter(prefix="/schedules", tags=["排期管理"])


class ResourceOption(BaseModel):
    id: int
    name: str
    department_name: Optional[str] = None
    job_title: Optional[str] = None
    role_names: list[str] = []


class PersonTreeNode(BaseModel):
    value: int | str
    label: str
    disabled: bool = False
    is_person: bool = False
    children: list["PersonTreeNode"] = Field(default_factory=list)


PersonTreeNode.model_rebuild()


@router.get("/stats", response_model=ScheduleStatsOut, summary="排期统计")
def stats(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> ScheduleStatsOut:
    return ScheduleStatsOut(**schedule_service.schedule_stats(db, current_user))


@router.get("/options/resources", response_model=list[ResourceOption], summary="资源人员选项")
def resource_options(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
    resource_type: Optional[str] = Query(
        None,
        description="可选；仅把匹配组织角色的人排到前面：instructor/streamer/shooting_edit/other",
    ),
) -> list[ResourceOption]:
    _ = current_user
    return [
        ResourceOption(**x)
        for x in schedule_service.list_resource_options(db, resource_type=resource_type)
    ]


@router.get(
    "/options/person-tree",
    response_model=list[PersonTreeNode],
    summary="组织架构人员树（部门下挂人）",
)
def person_tree(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> list[PersonTreeNode]:
    _ = current_user
    return [PersonTreeNode.model_validate(x) for x in schedule_service.list_person_tree(db)]


@router.get("/resource-load", response_model=ResourceLoadListOut, summary="资源负载")
def resource_load(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
    resource_type: str = Query(..., description="instructor/streamer/shooting_edit/other"),
    date_from: datetime = Query(...),
    date_to: datetime = Query(...),
) -> ResourceLoadListOut:
    items = schedule_service.resource_load(
        db,
        current_user,
        resource_type=resource_type,
        date_from=date_from,
        date_to=date_to,
    )
    return ResourceLoadListOut(items=[ResourceLoadOut(**x) for x in items])


@router.get("", response_model=ScheduleListOut, summary="排期列表")
def list_schedules(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
    status: Optional[str] = None,
    resource_type: Optional[str] = None,
    employee_id: Optional[int] = None,
    project_id: Optional[int] = None,
    project_task_id: Optional[int] = None,
    date_from: Optional[datetime] = None,
    date_to: Optional[datetime] = None,
    scope: Optional[str] = Query(None, description="mine/all"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
) -> ScheduleListOut:
    total, items = schedule_service.list_schedules(
        db,
        current_user,
        status=status,
        resource_type=resource_type,
        employee_id=employee_id,
        project_id=project_id,
        project_task_id=project_task_id,
        date_from=date_from,
        date_to=date_to,
        scope_filter=scope,
        page=page,
        page_size=page_size,
    )
    return ScheduleListOut(total=total, items=[ScheduleOut.model_validate(x) for x in items])


@router.post("", response_model=ScheduleOut, summary="申请排期")
def create_schedule(
    payload: ScheduleCreate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> ScheduleOut:
    return ScheduleOut.model_validate(schedule_service.create_schedule(db, current_user, payload))


@router.get("/{schedule_id}", response_model=ScheduleOut, summary="排期详情")
def get_schedule(
    schedule_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> ScheduleOut:
    return ScheduleOut.model_validate(schedule_service.get_schedule(db, current_user, schedule_id))


@router.patch("/{schedule_id}", response_model=ScheduleOut, summary="编辑排期")
def update_schedule(
    schedule_id: int,
    payload: ScheduleUpdate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> ScheduleOut:
    return ScheduleOut.model_validate(
        schedule_service.update_schedule(db, current_user, schedule_id, payload)
    )


@router.post("/{schedule_id}/confirm", response_model=ScheduleOut, summary="确认排期")
def confirm_schedule(
    schedule_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> ScheduleOut:
    return ScheduleOut.model_validate(
        schedule_service.confirm_schedule(db, current_user, schedule_id)
    )


@router.post("/{schedule_id}/coordinate", response_model=ScheduleOut, summary="请求协调")
def coordinate_schedule(
    schedule_id: int,
    payload: ScheduleCoordinateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> ScheduleOut:
    return ScheduleOut.model_validate(
        schedule_service.request_coordination(db, current_user, schedule_id, payload)
    )


@router.post("/{schedule_id}/start", response_model=ScheduleOut, summary="开始执行")
def start_schedule(
    schedule_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> ScheduleOut:
    return ScheduleOut.model_validate(
        schedule_service.start_schedule(db, current_user, schedule_id)
    )


@router.post("/{schedule_id}/complete", response_model=ScheduleOut, summary="完成排期")
def complete_schedule(
    schedule_id: int,
    payload: ScheduleCompleteRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> ScheduleOut:
    return ScheduleOut.model_validate(
        schedule_service.complete_schedule(db, current_user, schedule_id, payload)
    )


@router.post("/{schedule_id}/cancel", response_model=ScheduleOut, summary="取消排期")
def cancel_schedule(
    schedule_id: int,
    payload: ScheduleCancelRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> ScheduleOut:
    return ScheduleOut.model_validate(
        schedule_service.cancel_schedule(db, current_user, schedule_id, payload)
    )


@router.delete("/{schedule_id}", summary="删除排期（仅管理员，已完成也可删）")
def delete_schedule(
    schedule_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["schedule:view"]))],
) -> dict:
    schedule_service.delete_schedule(db, current_user, schedule_id)
    return {"ok": True, "message": "已删除"}
